### PR TITLE
Icacallbacks undelegation

### PR DESCRIPTION
**Pull request by sampocs**
Closes: TEST-164

## What is the purpose of the change

* The purpose of the PR is to address a bug which can cause the redemption rate to spike dramatically in both directions:
1) In the current implementation, stTokens are burned as soon as `msg_redeem_stake` is called, but before those tokens have actually unbonded - causing the redemption rate to be artificially high (↓ stAsset,  ↑ redemption rate )
2) Then if more liquid stakes occur, the number of stTokens minted grows larger, and is followed by the materialization of the unbonding which causes the redemption rate to drop close to 0 (↑stAsset, ↓staked balance, ↓redemption rate )
* The solution is to wait to burn the stTokens until after the native tokens have been unbonded from the host (done with the help of Aidan's beautiful callbacks module)

## Brief Changelog
  - Updated `EpochUnbondingRecords` proto to use the epoch number as an ID 
  - Updated `HostZoneUnbondings` proto (within `EpochUnbondingRecords`) to have both the native and stToken amount (so we know how many stToken to burn after acknowledgement)
  - Added undelegation callback 
  - Moved stToken burn from redeem to the undelegation callback

#### Context 
* An upgrade is proposed through governance at a specified block height
* The upgrade module checks the current block height in EndBlocker, and if the current height matches the height of an upgrade, it panics
* The chain is restarted with the new binary
* The upgrade module sees that the chain is being started at the same height that an upgrade occurred, so it grabs the name of the upgrade (that was passed from the proposal) and routes to the upgrade handler before starting the app

#### Upgrade Handlers
* `setupUpgradeHandlers()` is defined in `app/upgrades.go` and is called from `app.go`. It defines how each upgrade should be handled by essentially specifying a mapping from upgrade name → upgrade function. Each upgrade is registered here:
```
# app/upgrades.go
func setupUpgradeHandlers() {
      # set upgrade handler for v2
      # set upgrade handler for v3
      # etc...
}
``` 
* Then `app/upgrades/{upgradeName}/upgrades.go`, stores the specific logic for a given upgrade
* The main function here is `CreateUpgradeHandler` function (which is called from `setupUpgradeHandler` when it creates  the mapping of upgrade name → upgrade function)
* This function takes a VersionMap as a parameter which defines the version that each **module** should be on
* For simple upgrades with no state changes, nothing needs to take place here, and the original VersionMap should be returned
* However, if the **module** has state changes, you have to specify the change in consensus version in the [module](https://github.com/Stride-Labs/stride/pull/171/files#diff-b53d1eca824eaecf2eaaf3b0e446efe8af36ee2782a51d83903a832f758b9d0cR176) itself, and in the upgrade handler as follows:
```
# the 2 here is defined as the ConsensusVersion in module.go 
vm[recordstypes.ModuleName] = 2 
```
* When this upgrade handler is called, it will see that the `records` module needs to be migrated from version 1 to 2 and will run the corresponding migration

#### Migration 
* The migration must be registered under `RegisterServices` in `module.go` 
* This is similar to the [mapping](https://github.com/Stride-Labs/stride/pull/171/files#diff-b53d1eca824eaecf2eaaf3b0e446efe8af36ee2782a51d83903a832f758b9d0cR149) above in that it defines a migration function for each version change (e.g. version 1 to 2 → migration function). These migration functions are all stored in the same file: `{moduleName}/keeper/migrations.go`
```
func Migrate1to2() {}
func Migrate2to3() {}
...
```
* FINALLY, the actual state changes are defined under `{moduleName}/keeper/migrations/{upgradeName}/migrations.go`
* Since it is often the case that an upgrade will have to deserialize from an old proto type, those old types are stored in `{moduleName}/keeper/migrations/{previousUpgradeVersion}/types`
* **That last part is the most important part when it comes to getting ahead of these upgrades. When changing proto types, we should place a copy of the old types (that were changed) in `{moduleName}/keeper/migrations/{previousUpgradeVersion}/types`, and at the very least fill in the boilerplate under `{moduleName}/keeper/migrations/{upgradeName}/migrations.go`**

#### Summary
* `app/upgrades.go`: defines mapping of upgrade name → upgrade function
* `app/upgrades/{upgradeName}/upgrades.go`: defines which modules need to be upgraded
* `{moduleName}/keeper/migrations.go`: defines mapping of version change → migration function
* `{moduleName}/keeper/migrations/{upgradeName}/migrations.go`: migration logic (e.g. state changes) 
* `{moduleName}/keeper/migrations/{previousUpgradeVersion}/types`: old proto types 

## Testing and Verifying
- [X] This change should alters the behavior of the existing unit tests, but they have been updated accordingly
- [X] The simulation test (firehose) should show a stable redemption rate
```
# QRR:    Queried Redemption Rate
# CRR:    Calculated Redemption Rate (using each component)
# Stk:    Staked Balance
# UD:     Undelegated balance
# Module: Module Balance 
# st:     Supply of stATOM

# BEFORE THIS FIX (notice redemption rate jumps up then down)
H:104 | E:14 | QRR: 1.000000 | CRR: 1.000000 | Stk: 0        | UD: 0        | Module: 0        | st: 0
H:123 | E:16 | QRR: 1.000000 | CRR: 1.000000 | Stk: 105172   | UD: 113340   | Module: 150487   | st: 368999
H:133 | E:18 | QRR: 1.279709 | CRR: 1.279709 | Stk: 218512   | UD: 150487   | Module: 0        | st: 288346
H:148 | E:20 | QRR: 2.421299 | CRR: 2.421299 | Stk: 214556   | UD: 51       | Module: 0        | st: 88633
H:164 | E:22 | QRR: 0.402028 | CRR: 0.402028 | Stk: 35609    | UD: 24       | Module: 0        | st: 88633

# AFTER THIS FIX (notice redemption rate increases steadily and monotonically)
H:106 | E:14 | QRR: 1.000000 | CRR: 1.000000 | Stk: 0        | UD: 0        | Module: 0        | st: 0
H:125 | E:17 | QRR: 1.000000 | CRR: 1.000000 | Stk: 71451    | UD: 106306   | Module: 191242   | st: 368999
H:137 | E:18 | QRR: 1.000000 | CRR: 1.000000 | Stk: 330491   | UD: 38508    | Module: 0        | st: 368999
H:148 | E:20 | QRR: 1.000193 | CRR: 1.000193 | Stk: 264226   | UD: 51       | Module: 0        | st: 264226
H:164 | E:22 | QRR: 1.000896 | CRR: 1.000896 | Stk: 99357    | UD: 29       | Module: 0        | st: 99297
H:200 | E:26 | QRR: 1.001591 | CRR: 1.001591 | Stk: 99455    | UD: 11       | Module: 0        | st: 99297

Firehose Test Results:
2022/08/12 12:55:38 TOTAL: 100/100(100.00) Success
```
- [X] Integration tests
```
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC] host zone successfully registered
 ✓ [INTEGRATION-BASIC] liquid stake mints stATOM
 ✓ [INTEGRATION-BASIC] tokens were transferred to GAIA after liquid staking
 ✓ [INTEGRATION-BASIC] tokens on GAIA were staked
 ✓ [INTEGRATION-BASIC] redemption works
 ✓ [INTEGRATION-BASIC] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC] rewards are being reinvested (delegated balance increasing)
 ✓ [INTEGRATION-BASIC] exchange rate is updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct
```
## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no) no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no) no
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  / README.md  /   not documented) not documented

### CODE REVIEW

**Comment by asalzmann** [comment]
just a heads up - seems like some migration-related changes got rolled into this!

**Comment by sampocs** [comment]
I added that specifically for this PR! I had to change the proto types and figured it made sense to get ahead of the upgrade rather than trying to figure it out last minute

**Comment by asalzmann** [comment]
I merged `icacallbacks`, so you should be able to compare this branch to `main` now

**Comment by sampocs** [comment]
This is ready for review! However, I want to re-run firehose before merging (which I wont get to until tomorrow)

**Comment by sampocs** [comment]
Firehose test passed - results added to PR description

**Comment by asalzmann** [comment]
@sampocs taking a look now - can you add a brief description of the upgrade code?

**Comment by sampocs** [comment]
> Can you walk through this example in more detail?

I'm very glad you asked this cause I actually caught a mistake in the way I was thinking about it. Sorry for the confusion, this is the right way to think about the issue:

To give a simple example, let's say there's originally 100 ATOM stake, someone redeems 50 stATOM, and then someone else redeems another 25 ATOM...

**Old Scenario** 
```
1. Initially we have:
Staked Balance:      100    |  stATOM: 100   |     Redemption Rate: 1.00

2. Then as soon as the redemption is triggered (50 stATOM for 50 ATOM), the stATOM are burned immediately
However, at this point, we haven't yet decremented the staked balance because we haven't initiated the unbonding or received the ack yet
Staked Balance:      100   |  stATOM: 50    |     Redemption Rate: 2.0

3. Now lets say there's another redemption of 25 stATOM, this time for 50 ATOM (because the RR has changed)
However, the staked balance has still not been touched
Staked Balance:      100   |  stATOM: 25   |     Redemption Rate: 4.00

5. Finally, once the acks comes back after the unbondings, the staked balance will decrease by 100 (50 from each of the two redemptions)
And the redemption rate will now be 0
Staked Balance:       0   |   stATOM: 25    |     Redemption Rate: 0.0 
```

The root of this issue surfaces from a lack of atomicity between the burning of stTokens and the corresponding stakedBal decrement. The solution is to apply these two operations at the same time - which is in the undelegation callback

**Comment by asalzmann** [comment]
This makes a lot of sense, thank you for walking through this in detail! That's a tricky edge case

**Comment by sampocs** [comment]
Just updated the description with more details around the upgrade!

File: x/stakeibc/keeper/unbonding_records.go [hunk 1]
```diff
@@ -111,9 +117,25 @@ func (k Keeper) SendHostZoneUnbondings(ctx sdk.Context, hostZone types.HostZone)
...
 	}
-	// now we have to handle the overflow amount
-	_, err = k.SubmitTxsDayEpoch(ctx, hostZone.GetConnectionId(), msgs, *delegationAccount)
+	marshalledCallbackArgs, err := k.MarshalUndelegateCallbackArgs(ctx, undelegateCallback)
+	if err != nil {
+		k.Logger(ctx).Error(err.Error())
+		return false
+	}
+
+	_, err = k.SubmitTxsDayEpoch(ctx, hostZone.GetConnectionId(), msgs, *delegationAccount, "undelegate", marshalledCallbackArgs)
```
Comment (asalzmann) [inline x/stakeibc/keeper/unbonding_records.go:138]: note: something we should test is unbonding from ~50 validators in a single transaction on mainnet (we can delegate to the top 50 validators, then issue an unbonding tx with all of the messages)

**Review by asalzmann — CHANGES_REQUESTED** [review]
This is awesome! Much more efficient and easier to reason about, thank you for making these changes. I requested a couple of small changes but the callback logic looks good overall. I haven't reviewed the migration code yet, going to do that next!

File: proto/records/genesis.proto [hunk 2]
```diff
@@ -83,23 +83,18 @@ message HostZoneUnbonding {
     // transfer success
     TRANSFERRED = 3;
   }
-  uint64 amount = 1;
-  string denom = 2;
-  string hostZoneId = 3;
+  uint64 stTokenAmount = 1;
```
[the remark this answers is not in the document]
Comment (sampocs) [reply proto/records/genesis.proto:86]: Good callout! We should probably get in the habit of generating fresh typescrypt protos when ever we deploy an upgrade

File: x/records/module.go [hunk 3]
```diff
@@ -168,7 +173,7 @@ func (am AppModule) ExportGenesis(ctx sdk.Context, cdc codec.JSONCodec) json.Raw
 }
 
 // ConsensusVersion implements ConsensusVersion.
-func (AppModule) ConsensusVersion() uint64 { return 1 }
+func (AppModule) ConsensusVersion() uint64 { return 2 }
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/records/module.go:176]: it's needed when we upgrade the types for a module. Just added a more thorough description in the PR overview

File: x/records/types/genesis.go [hunk 4]
```diff
@@ -38,18 +37,6 @@ func (gs GenesisState) Validate() error {
 		}
 		userRedemptionRecordIdMap[elem.Id] = true
 	}
-	// Check for duplicated ID in epochUnbondingRecord
-	epochUnbondingRecordIdMap := make(map[uint64]bool)
-	epochUnbondingRecordCount := gs.GetEpochUnbondingRecordCount()
-	for _, elem := range gs.EpochUnbondingRecordList {
-		if _, ok := epochUnbondingRecordIdMap[elem.Id]; ok {
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/records/types/genesis.go:45]: Ah good point. I was thinking we would never create unbonding records at genesis but I guess we would if we were ever doing a hard fork right?

File: x/stakeibc/keeper/icacallbacks.go [hunk 5]
```diff
@@ -46,14 +47,11 @@ func (c ICACallbacks) AddICACallback(id string, fn interface{}) icacallbackstype
 }
 
 func (c ICACallbacks) RegisterICACallbacks() icacallbackstypes.ICACallbackHandler {
-	a := c.AddICACallback("delegate", ICACallback(DelegateCallback))
-	return a.(ICACallbacks)
+	c.AddICACallback("delegate", ICACallback(DelegateCallback))
+	c.AddICACallback("undelegate", ICACallback(UndelegateCallback))
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icacallbacks.go:51]: done

Also I'm noticing now that it does seem a little odd to have an instance of the object returned by a function in the object itself? Is this a design pattern you've seen before?

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 6]
```diff
@@ -94,7 +94,7 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
 	if !found {
 		return nil, sdkerrors.Wrapf(types.ErrInvalidHostZone, "host zone not found in unbondings: %s", hostZone.ChainId)
 	}
-	hostZoneUnbonding.Amount += inCoin.Amount.Uint64()
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:97]: gladly!

I know I've double checked on this before, but something still isn't adding up. Can you clarify this for me? It obvioulsy works so I must be missing something 

```
// (I'm using old code without renaming)
coinDenom := types.StAssetDenomFromHostZoneDenom(hostZone.HostDenom) // returns stuatom
nativeAmount := sdk.NewDec(msg.Amount).Mul(hostZone.RedemptionRate)  // returns amount to be redeemed on the host zone
coinString := nativeAmount.RoundInt().String() + coinDenom  // Creates a coin with native amount but st denom??
// shouldnt this be the host denom instead?
```

File: x/stakeibc/keeper/msg_server_redeem_stake.go [hunk 7]
```diff
@@ -105,12 +105,8 @@ func (k msgServer) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake)
...
 	}
 
-	// burn stAssets upon successful unbonding
-	err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, redeemCoin)
-	if err != nil {
-		k.Logger(ctx).Error(fmt.Sprintf("Failed to burn stAssets upon successful unbonding %v", err))
-		return nil, sdkerrors.Wrapf(types.ErrInsufficientFunds, "couldn't burn %d %s tokens in module account. err: %s", msg.Amount, coinDenom, err.Error())
-	}
+	// record the number of stAssets that should be burned after unbonding
+	hostZoneUnbonding.StTokenAmount += uint64(msg.Amount)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:109]: done

File: x/stakeibc/keeper/icacallbacks_undelegate.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	k.Logger(ctx).Info(fmt.Sprintf("UndelegateCallback, HostZone: %s", undelegateCallback.HostZoneId))
+
+	hostZone, found := k.GetHostZone(ctx, undelegateCallback.HostZoneId)
+	if !found {
+		return sdkerrors.Wrapf(sdkerrors.ErrKeyNotFound, "Host zone not found: %s", undelegateCallback.HostZoneId)
+	}
+
+	// Redelegate to each validator and updated host zone staked balance if successful
+	for _, undelegation := range undelegateCallback.SplitDelegations {
+		undelegateAmt := int64(undelegation.Amount)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icacallbacks_undelegate.go:62]: done

File: x/stakeibc/keeper/icacallbacks_undelegate.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+		hostZoneUnbondings.UnbondingTime = cast.ToUint64(latestCompletionTime.UnixNano())
+		k.RecordsKeeper.SetEpochUnbondingRecord(ctx, epochUnbondingRecord)
+
+		logMsg := fmt.Sprintf("Set unbonding time to %s for host zone %s's unbonding record: %d",
+			latestCompletionTime.String(), hostZone.ChainId, epochNumber)
+		k.Logger(ctx).Info(logMsg)
+	}
+
+	// Burn stTokens
+	stCoinDenom := "st" + hostZone.HostDenom
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icacallbacks_undelegate.go:131]: done

File: x/stakeibc/keeper/icacallbacks_undelegate.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "could not parse burnCoin: %s. err: %s", stCoinString, err.Error())
+	}
+	err = k.bankKeeper.BurnCoins(ctx, types.ModuleName, sdk.NewCoins(stCoin))
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("Failed to burn stAssets upon successful unbonding %s", err.Error()))
+		return sdkerrors.Wrapf(types.ErrInsufficientFunds, "couldn't burn %d %s tokens in module account. err: %s", stAmountToBurn, stCoinDenom, err.Error())
+	}
+
+	k.Logger(ctx).Info(fmt.Sprintf("Total supply %s", k.bankKeeper.GetSupply(ctx, "stuatom")))
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icacallbacks_undelegate.go:143]: good eye!

File: x/stakeibc/keeper/icacallbacks_undelegate.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+		}
+		hostZoneUnbondings, found := epochUnbondingRecord.HostZoneUnbondings[hostZone.ChainId]
+		if !found {
+			errMsg := fmt.Sprintf("Host zone not found (%s) in epoch unbonding record: %d", hostZone.ChainId, epochNumber)
+			k.Logger(ctx).Error(errMsg)
+			return sdkerrors.Wrapf(sdkerrors.ErrKeyNotFound, errMsg)
+		}
+
+		// Keep track of the stTokens that need to be burned
+		stAmountToBurn += int64(hostZoneUnbondings.StTokenAmount)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icacallbacks_undelegate.go:118]: done

File: x/stakeibc/keeper/icacallbacks_undelegate.go [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+	// Update the status and time of each unbonding record and grab the number of stTokens that need to be burned
+	stAmountToBurn := int64(0)
+	for _, epochNumber := range undelegateCallback.UnbondingEpochNumbers {
+		epochUnbondingRecord, found := k.RecordsKeeper.GetEpochUnbondingRecord(ctx, epochNumber)
+		if !found {
+			errMsg := fmt.Sprintf("Unable to find epoch unbonding record for epoch: %d", epochNumber)
+			k.Logger(ctx).Error(errMsg)
+			return sdkerrors.Wrapf(sdkerrors.ErrKeyNotFound, errMsg)
+		}
+		hostZoneUnbondings, found := epochUnbondingRecord.HostZoneUnbondings[hostZone.ChainId]
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icacallbacks_undelegate.go:110]: done

File: x/stakeibc/keeper/unbonding_records.go (hunk 1 above)
Comment (sampocs) [reply x/stakeibc/keeper/unbonding_records.go:138]: good call, added as a todo in the Launch Plan

File: x/stakeibc/keeper/icacallbacks.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (asalzmann) [reply x/stakeibc/keeper/icacallbacks.go]: the IBC router uses it as well - it is a bit of a weird pattern

File: x/records/types/genesis.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (asalzmann) [reply x/records/types/genesis.go:45]: yep I think we'd have to in that case!

File: x/stakeibc/keeper/msg_server_redeem_stake.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (asalzmann) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:97]: good catch! I think this is a bug, but the Denom in question isn't actually used anywhere
- we create the UserRedemptionRecord with the denom pulled off of the host zone
- we use inCoin.Amount to add to hostZoneUnbonding
hopefully, the naming prevents us from making similar errors in the future, it's lucky we weren't using the denom

[the remark this answers is not in the document]
Comment (asalzmann) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:97]: do you mind making this update as well?

[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/msg_server_redeem_stake.go:97]: yup, pushing it in a sec

**Comment by asalzmann** [comment]
pending
- @vish-stride taking a look at this https://github.com/Stride-Labs/stride/pull/171#discussion_r944823398
- updating `inCoin` -> `nativeCoin` and fixing the denom error

lgtm!

**Comment by sampocs** [comment]
integration tests still pass

File: x/stakeibc/keeper/ibc_handlers.go [hunk 13]
```diff
@@ -259,62 +241,3 @@ func (k *Keeper) HandleSend(ctx sdk.Context, msg sdk.Msg, sequence string) error
...
-	}
-	zone.StakedBal -= undelegateAmt
-	k.SetHostZone(ctx, *zone)
-
-	epochTracker, found := k.GetEpochTracker(ctx, "day")
-	if !found {
-		return sdkerrors.Wrapf(types.ErrInvalidLengthEpochTracker, "no number for epoch (%s)", "day")
-	}
-	currentEpochNumber := epochTracker.GetEpochNumber()
-	for _, epochUnbonding := range k.RecordsKeeper.GetAllEpochUnbondingRecord(ctx) {
```
[the remark this answers is not in the document]
Comment (vish-stride) [reply x/stakeibc/keeper/ibc_handlers.go:299]: I think that's right, but we didn't assume they're processed every batch, only until the callback came back as a success iirc

**Review by asalzmann — APPROVED** [review]
lgtm!

**Pull request merged**